### DISCUSSION TITLE
Orno or-we-517 connection to ESP32 problem

### DISCUSSION THREAD

**Discussion by HaraldNoack** [arendst/Tasmota#17545] [Q&A]
Hello,

I compiled the source 12.1.1 with the TasmotaCompiler an the following switches user_config_override.h:
``
#ifndef _USER_CONFIG_OVERRIDE_H_
#define _USER_CONFIG_OVERRIDE_H_

#ifdef CODE_IMAGE_STR
  #undef CODE_IMAGE_STR
#endif
#define CODE_IMAGE_STR "TasmoCompiler-esp32generic"

#ifdef STA_SSID1
  #undef STA_SSID1
#endif
#define STA_SSID1	"XXXX"

#ifdef STA_PASS1
  #undef STA_PASS1
#endif
#define STA_PASS1	"YYYY"

#ifdef USE_ENERGY_SENSOR
  #undef USE_ENERGY_SENSOR
#endif

#ifdef USE_SR04
  #undef USE_SR04
#endif

#ifdef USE_VL53L0X
  #undef USE_VL53L0X
#endif

#ifdef USE_HRXL
  #undef USE_HRXL
#endif

#ifdef USE_DYP
  #undef USE_DYP
#endif

#ifdef USE_VL53L1X
  #undef USE_VL53L1X
#endif

#ifdef USE_EMULATION
  #undef USE_EMULATION
#endif

#ifdef USE_EMULATION_HUE
  #undef USE_EMULATION_HUE
#endif

#ifdef USE_EMULATION_WEMO
  #undef USE_EMULATION_WEMO
#endif

#ifdef USE_ADC_VCC
  #undef USE_ADC_VCC
#endif

#ifdef USE_TASMOTA_CLIENT
  #undef USE_TASMOTA_CLIENT
#endif

#ifdef USE_BERRY
  #undef USE_BERRY
#endif
#define USE_BERRY

#ifdef USE_BERRY_PSRAM
  #undef USE_BERRY_PSRAM
#endif
#define USE_BERRY_PSRAM

#ifdef USE_BLE_ESP32
  #undef USE_BLE_ESP32
#endif

#ifdef USE_MI_ESP32
  #undef USE_MI_ESP32
#endif

#ifdef USE_DISPLAY
  #undef USE_DISPLAY
#endif

#ifdef USE_DISPLAY_MODES1TO5
  #undef USE_DISPLAY_MODES1TO5
#endif

#ifdef USE_DISPLAY_LCD
  #undef USE_DISPLAY_LCD
#endif

#ifdef USE_DISPLAY_SSD1306
  #undef USE_DISPLAY_SSD1306
#endif

#ifdef USE_DISPLAY_MATRIX
  #undef USE_DISPLAY_MATRIX
#endif

#ifdef USE_DISPLAY_ILI9341
  #undef USE_DISPLAY_ILI9341
#endif

#ifdef USE_DISPLAY_EPAPER_29
  #undef USE_DISPLAY_EPAPER_29
#endif

#ifdef USE_DISPLAY_EPAPER_42
  #undef USE_DISPLAY_EPAPER_42
#endif

#ifdef USE_DISPLAY_SH1106
  #undef USE_DISPLAY_SH1106
#endif

#ifdef USE_DISPLAY_ILI9488
  #undef USE_DISPLAY_ILI9488
#endif

#ifdef USE_DISPLAY_SSD1351
  #undef USE_DISPLAY_SSD1351
#endif

#ifdef USE_DISPLAY_RA8876
  #undef USE_DISPLAY_RA8876
#endif

#ifdef USE_DISPLAY_SEVENSEG
  #undef USE_DISPLAY_SEVENSEG
#endif

#ifdef USE_DISPLAY_ST7789
  #undef USE_DISPLAY_ST7789
#endif

#ifdef USE_DISPLAY_SSD1331
  #undef USE_DISPLAY_SSD1331
#endif

#ifdef USE_DOMOTICZ
  #undef USE_DOMOTICZ
#endif

#ifdef USE_ETHERNET
  #undef USE_ETHERNET
#endif

#ifdef USE_HOME_ASSISTANT
  #undef USE_HOME_ASSISTANT
#endif

#ifdef USE_MCP230xx
  #undef USE_MCP230xx
#endif

#ifdef USE_MCP230xx_OUTPUT
  #undef USE_MCP230xx_OUTPUT
#endif

#ifdef USE_MCP230xx_DISPLAYOUTPUT
  #undef USE_MCP230xx_DISPLAYOUTPUT
#endif

#ifdef USE_I2C
  #undef USE_I2C
#endif

#ifdef USE_IR_REMOTE
  #undef USE_IR_REMOTE
#endif

#ifdef USE_IR_REMOTE_FULL
  #undef USE_IR_REMOTE_FULL
#endif

#ifdef USE_KNX
  #undef USE_KNX
#endif

#ifdef USE_BH1750
  #undef USE_BH1750
#endif

#ifdef USE_VEML6070
  #undef USE_VEML6070
#endif

#ifdef USE_TSL2561
  #undef USE_TSL2561
#endif

#ifdef USE_SI1145
  #undef USE_SI1145
#endif

#ifdef USE_APDS9960
  #undef USE_APDS9960
#endif

#ifdef USE_VEML6075
  #undef USE_VEML6075
#endif

#ifdef USE_MAX44009
  #undef USE_MAX44009
#endif

#ifdef USE_TSL2591
  #undef USE_TSL2591
#endif

#ifdef USE_AS3935
  #undef USE_AS3935
#endif

#ifdef USE_VEML7700
  #undef USE_VEML7700
#endif

#ifdef USE_MHZ19
  #undef USE_MHZ19
#endif

#ifdef USE_SENSEAIR
  #undef USE_SENSEAIR
#endif

#ifdef USE_PMS5003
  #undef USE_PMS5003
#endif

#ifdef USE_MGS
  #undef USE_MGS
#endif

#ifdef USE_NOVA_SDS
  #undef USE_NOVA_SDS
#endif

#ifdef USE_SGP30
  #undef USE_SGP30
#endif

#ifdef USE_CCS811
  #undef USE_CCS811
#endif

#ifdef USE_SCD30
  #undef USE_SCD30
#endif

#ifdef USE_SPS30
  #undef USE_SPS30
#endif

#ifdef USE_HPMA
  #undef USE_HPMA
#endif

#ifdef USE_IAQ
  #undef USE_IAQ
#endif

#ifdef USE_T67XX
  #undef USE_T67XX
#endif

#ifdef USE_VINDRIKTNING
  #undef USE_VINDRIKTNING
#endif

#ifdef USE_SCD40
  #undef USE_SCD40
#endif

#ifdef USE_HM330X
  #undef USE_HM330X
#endif

#ifdef USE_DISCOVERY
  #undef USE_DISCOVERY
#endif

#ifdef USE_MQTT_TLS
  #undef USE_MQTT_TLS
#endif

#ifdef USE_RULES
  #undef USE_RULES
#endif

#ifdef USE_EXPRESSION
  #undef USE_EXPRESSION
#endif

#ifdef SUPPORT_IF_STATEMENT
  #undef SUPPORT_IF_STATEMENT
#endif

#ifdef USE_RC_SWITCH
  #undef USE_RC_SWITCH
#endif

#ifdef USE_UFILESYS
  #undef USE_UFILESYS
#endif
#define USE_UFILESYS

#ifdef USE_SDCARD
  #undef USE_SDCARD
#endif
#define USE_SDCARD

#ifdef GUI_TRASH_FILE
  #undef GUI_TRASH_FILE
#endif
#define GUI_TRASH_FILE

#ifdef GUI_EDIT_FILE
  #undef GUI_EDIT_FILE
#endif
#define GUI_EDIT_FILE

#ifdef USE_SCRIPT
  #undef USE_SCRIPT
#endif
#define USE_SCRIPT

#ifdef USE_SPI
  #undef USE_SPI
#endif
#define USE_SPI

#ifdef USE_HLW8012
  #undef USE_HLW8012
#endif

#ifdef USE_CSE7766
  #undef USE_CSE7766
#endif

#ifdef USE_PZEM004T
  #undef USE_PZEM004T
#endif

#ifdef USE_MCP39F501
  #undef USE_MCP39F501
#endif

#ifdef USE_PZEM_AC
  #undef USE_PZEM_AC
#endif

#ifdef USE_PZEM_DC
  #undef USE_PZEM_DC
#endif

#ifdef USE_ADE7953
  #undef USE_ADE7953
#endif

#ifdef USE_SDM120
  #undef USE_SDM120
#endif

#ifdef USE_DDS2382
  #undef USE_DDS2382
#endif

#ifdef USE_SDM630
  #undef USE_SDM630
#endif

#ifdef USE_DDSU666
  #undef USE_DDSU666
#endif

#ifdef USE_SOLAX_X1
  #undef USE_SOLAX_X1
#endif

#ifdef USE_LE01MR
  #undef USE_LE01MR
#endif

#ifdef USE_BL09XX
  #undef USE_BL09XX
#endif

#ifdef USE_TELEINFO
  #undef USE_TELEINFO
#endif

#ifdef USE_IEM3000
  #undef USE_IEM3000
#endif

#ifdef USE_WE517
  #undef USE_WE517
#endif

#ifdef USE_ENERGY_DUMMY
  #undef USE_ENERGY_DUMMY
#endif

#ifdef USE_SONOFF_SC
  #undef USE_SONOFF_SC
#endif

#ifdef USE_DS18x20
  #undef USE_DS18x20
#endif

#ifdef USE_DHT
  #undef USE_DHT
#endif

#ifdef USE_SHT
  #undef USE_SHT
#endif

#ifdef USE_HTU
  #undef USE_HTU
#endif

#ifdef USE_BMP
  #undef USE_BMP
#endif

#ifdef USE_SHT3X
  #undef USE_SHT3X
#endif

#ifdef USE_LM75AD
  #undef USE_LM75AD
#endif

#ifdef USE_AZ7798
  #undef USE_AZ7798
#endif

#ifdef USE_MAX31855
  #undef USE_MAX31855
#endif

#ifdef USE_MLX90614
  #undef USE_MLX90614
#endif

#ifdef USE_MAX31865
  #undef USE_MAX31865
#endif

#ifdef USE_HIH6
  #undef USE_HIH6
#endif

#ifdef USE_DHT12
  #undef USE_DHT12
#endif

#ifdef USE_DS1624
  #undef USE_DS1624
#endif

#ifdef USE_AHT1x
  #undef USE_AHT1x
#endif

#ifdef USE_HDC1080
  #undef USE_HDC1080
#endif

#ifdef USE_MCP9808
  #undef USE_MCP9808
#endif

#ifdef USE_HP303B
  #undef USE_HP303B
#endif

#ifdef USE_LMT01
  #undef USE_LMT01
#endif

#ifdef USE_AM2320
  #undef USE_AM2320
#endif

#ifdef USE_TIMERS
  #undef USE_TIMERS
#endif
#define USE_TIMERS

#ifdef USE_TUYA_MCU
  #undef USE_TUYA_MCU
#endif

#ifdef USE_WEBSERVER
  #undef USE_WEBSERVER
#endif
#define USE_WEBSERVER

#ifdef USE_WS2812
  #undef USE_WS2812
#endif

#ifdef USE_ZIGBEE
  #undef USE_ZIGBEE
#endif

#ifdef MY_LANGUAGE
  #undef MY_LANGUAGE
#endif
#define MY_LANGUAGE	en_GB

#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif

#ifndef USE_AUTOCONF
#define USE_AUTOCONF
#endif

#ifndef USE_ENERGY_SENSOR
#define USE_ENERGY_SENSOR
#endif

#ifndef USE_WE517
#define USE_WE517
#endif
#endif
``

Then I flashed it and inserted the following script for 2 S0-Counter on GPIO12 and GPIO14 1000Pulse/kWh.

``
>D 
>B
>M 2
+1,12,c,1,-10,Heizung
1,1-0:1.8.0*255(@1000, Energie,KWh,Total_in,3)
1,1-0:1.7.0*255(@0.01667, Leistung,W,Power_actual,0)
+2,14,c,1,-10,WWasser
2,1-0:1.8.0*255(@1000, Energie,KWh,Total_in,3)
2,1-0:1.7.0*255(@0.01667, Leistung,W,Power_actual,0)
#  
``

Tested it and it works:-)

Then I set the pins for the or-we-517 connected the ttl-r485 adapter and see the TX-Led flashing every second but no update in main screen :-(

![image](https://user-images.githubusercontent.com/57684656/210071244-fa520c59-7227-4ec1-8a7c-439107f812a6.png)
![image](https://user-images.githubusercontent.com/57684656/210071153-0aa12c7c-c303-46c8-9948-7dc8814b0402.png)

When I connect a USB-TTL adapter to the PC and connect it wit the TTL-RS485 adapter and use the ONRO PC software 9600 Baud 8E1 it works.

I tested RX-RX TX-TX and RX-TX TX-TX connection between ESP32 and TTL-RS485. Nothing works.
Is it a problem with the PIN-CORE-binding ? Core 0 or 1 ? There is no documentation.
Have I to use special pins?

THANKS for your HELP!!!!
Best Regards
Harald

**Comment by maximevince** [comment]
Mm, hard to tell what's wrong like that... Maybe you could try a different pair of pins? If you have a logic analyzer, you could probe both pins to "see" whats going on.
Are the voltage levels between the ESP32 and your TTL-RS485 adapter correct?

**Reply by HaraldNoack** [reply]
Thanks,

I played a little bit with my logic analyzer :-)

This are the measured signals direct at the USB-CP2102-module or the ESP32 Devkit C V4 module pins. 

![PC-Software_ORNO_9600_8E1_OK](https://user-images.githubusercontent.com/57684656/210184772-150b8d26-2a12-4b2f-a847-21bb2273a20d.png)
This is the data the PC sends to the we-517 
The UART-decoder is configured 9600Baud 8Bit Parity EVEN 1 Stop bit  (8E1)
NO ERRORS :-)

Then I changed to the TASMOTA programmed ESP32 Devkit C V4. 

WE517 TX is on GPIO23
![Tasmota_Config_1](https://user-images.githubusercontent.com/57684656/210184826-e2b0b342-55f8-4ef7-bf3d-7bbf02809a32.png)

![Tasmota_GPIO23_ORNO_9600_8E1_CRC_ERROR](https://user-images.githubusercontent.com/57684656/210184843-a68af44a-0626-4f8c-9443-b6c95b206670.png)
As we can see, there are Parity-ERRORS and so the we-517 ignores the data :-(

I changed the WE517 TX to GPIO17 (TX UART2)
![Tasmota_GPIO17_UART2_ORNO_9600_8E1_CRC_ERROR](https://user-images.githubusercontent.com/57684656/210184870-94c0867a-660f-430f-b321-4843873b9573.png)
The same Parity-ERRORS as before :-(

At last I changed the WE517 TX to GPIO1 (TX UART0)
![Tasmota_GPIO1_UART0_ORNO_9600_8E1_CRC_ERROR](https://user-images.githubusercontent.com/57684656/210184890-8f40a65e-a84d-4030-a700-cb4e8ebb3ff5.png)
The same Parity-ERRORS as before :-(

![Tasmota_Parity_ERROR_zoom](https://user-images.githubusercontent.com/57684656/210185779-1977c592-0ae4-49e1-8ce1-8a564074b72e.png)

Could it be that there is a bug in the parity-selection or HW-UART-driver ?

THANKS for your HELP!!!!
Best Regards 
Harald

**Reply by arendst** [reply]
Great job! As it stands it looks like the driver has never been tested with an esp32; serialconfig 8E1 is only provided to the default serial function which is probably not used by most esp32 configs as it has 1 or 2 additional serial functions.

I 'll fix it today and let you know when to test.

If you cannot wait you might want to try latest NRG modbus driver (which needs rules instead of script) default in development branch. With the correct, stripped down version, of the or-we517 interface like voltage only, you might be able to test it's communication with 8E1. Look into file nrg_29_modbus.ino how to use.

**Reply by arendst** [reply]
Added a fix to development branch as https://github.com/arendst/Tasmota/commit/cb6dd75d8b75f21472989c5498d46d0e1fff03b5

Pre-compiled available in one hour at http://ota.tasmota.com/tasmota32

**Reply by HaraldNoack** [reply]
Thanks for your help !!

please can you send  the link to the correct *.bin to flash and test it.

THANKS

**Reply by arendst** [reply]
Ah I see the driver is NOT enabled in the default binaries. You'll have to compile yourselves using the development branch.

**Reply by arendst** [reply]
I will enable it in the default tasmota32.bin binary. Again available in one hour at forementionend link.

**Reply by HaraldNoack** [reply]
Thanks !!!!!!

**Reply by HaraldNoack** [reply]
I tried the new tasmota32.bin and used GPIO22/23 for WE517RX/WE517TX bound to ESP32 devkit (0).

IT WORKS !!!!!! :-)

There is only one little problem:
The voltage displayed and sent via MQTT has no position after decimal point (235 instead of 235.2 displayed an the meter).
The frequency displayed and sent via MQTT has no position after decimal point (50 instead of 49.98 displayed an the meter).

Please can you implement: 
register 0x001C "Total Active Power"
register 0x0108 "Forward Active Energy"
register 0x0110 "Reverse Active Energy"

[OR-WE-517_MODBUS_Registers_List.pdf](https://github.com/arendst/Tasmota/files/10332304/OR-WE-517_MODBUS_Registers_List.pdf)

This 3 registers are important, if you want to make a zero power flow from your solar inverter into the grid or want to control the load current of your battery bank.

Please can you let the WE517 driver activated for the future and implement the corrected driver in the main branch. 

THANKS for your HELP!!!!
Best Regards 
Harald

**Reply by arendst** [reply]
> The voltage displayed and sent via MQTT has no position after decimal point (235 instead of 235.2 displayed an the meter).
The frequency displayed and sent via MQTT has no position after decimal point (50 instead of 49.98 displayed an the meter).

Read the docs. Look for commands ``Voltres 1`` and ``FreqRes 2``.

> Please can you implement:
register 0x001C "Total Active Power"
register 0x0108 "Forward Active Energy"
register 0x0110 "Reverse Active Energy"

To use these registers it's time you start using another driver. The generic modbus driver allows for users to configure any register they like. An example config looks like this:
```
rule3 on file#modbus do {"Name":"WE517","Baud":9600,"Config":"8E1","Address":1,"Function":3,"Voltage":[0xE,0x10,0x12],"Current":[0x16,0x18,0x1A],"Power":[0x1E,0x20,0x22],"ReactivePower":[0x26,0x28,0x2A],"Factor":[0x36,0x38,0x3A],"Frequency":0x14,"Total":0x100,"ExportActive":0x110,"User":[{"J":"ImportActive","G":"Import Active","R":0x108,"U":"kWh","D":24},{"J":"TotalPower","G":"Active Power Total","R":0x1C,"U":"W","D":23}]} endon
```

To use it change the current WE517 RX/TX GPIO's into NrgMbs RxTx GPIO's and execute command ``SerialBuffer 800``. The execute the above rule3 command as is.

**Reply by HaraldNoack** [reply]
Please is it possible, that you post the above posted rule as a script, because I need als some S0-Counter and disabled rules during compilation and I am a rule and script newbie.

After I finished I post a detailed description in "Show and tell".

THANKS !!!!!

**Reply by arendst** [reply]
Alas, currently it's only rules OR script.

I'll ask @gemu2015 if it's possible to provide the modbus file as JSON in his script environment but I fear it's kind of overkill as his script and sml files could do almost the same as the rule solution albeit without the generic ``VoltRes`` etc. commands.

**Reply by gemu2015** [reply]
it would really be an overkill. i will have a look if it could be made without too much effort.
otherwise you can easily build a descriptor for any Modbus device as ORNO and of course there is also a VoltRes equivalent.

**Reply by gemu2015** [reply]
but what about the same choices as with display descriptors for your modbus driver?

they can be 
1. a file
2. a section of script
3. a rule3 text
4. a #define in user_config_override.h

**Reply by arendst** [reply]
My comments:
1. The default ESP8266 binaries have no filesystem hence the use of the rules area while still allowing rules. I'll see if I can add it as an option on ESP32 hardware where a filesystem is always present.
2. Adding the modbus info with a new script command and being able to read it like the current rule solution would solve the issue.
3. is currently the only solution
4. would need recompiling which I don't like.

Thx for the ideas.

**Reply by gemu2015** [reply]
ok i checked the source and it will be easy to add RuleLoadFile equivalent to scripts.
i will however wait until you are ready with other modifications of this driver to avoid simultaneous editing of the same file.

**Reply by arendst** [reply]
Great. Let me add the file load part. Once done I let you know.

**Reply by arendst** [reply]
Latest dev (https://github.com/arendst/Tasmota/commit/18c38ec05d07662828bb03f85008ab6fa8390efe) provides support for filesystem too. Use a file called ``modbus.json`` with the following content:
```
{"Name":"WE517","Baud":9600,"Config":"8E1","Address":1,"Function":3,"Voltage":[0xE,0x10,0x12],"Current":[0x16,0x18,0x1A],"Power":[0x1E,0x20,0x22],"ReactivePower":[0x26,0x28,0x2A],"Factor":[0x36,0x38,0x3A],"Frequency":0x14,"Total":0x100,"ExportActive":0x110,"User":[{"J":"ImportActive","G":"Import Active","R":0x108,"U":"kWh","D":24},{"J":"TotalPower","G":"Active Power Total","R":0x1C,"U":"W","D":23}]}
```

**Reply by arendst** [reply]
So script is now also supported. You should add a line like this:
```
>y{"Name":"WE517","Baud":9600,"Config":"8E1","Address":1,"Function":3,"Voltage":[0xE,0x10,0x12],"Current":[0x16,0x18,0x1A],"Power":[0x1E,0x20,0x22],"ReactivePower":[0x26,0x28,0x2A],"Factor":[0x36,0x38,0x3A],"Frequency":0x14,"Total":0x100,"ExportActive":0x110,"User":[{"J":"ImportActive","G":"Import Active","R":0x108,"U":"kWh","D":24},{"J":"TotalPower","G":"Active Power Total","R":0x1C,"U":"W","D":23}]}#
```
